torch/_dynamo/utils.py

```diff
@@ -799,6 +805,105 @@ def get_compilation_metrics() -> List[Union[CompilationMetrics, BwdCompilationMe
     return list(_compilation_metrics)
 
 
+class ChromiumEventLogger:
+    """Logs chromium events to structured logs. tlparse will concatenate these into a perfetto UI link.
+
+    See https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.yr4qxyxotyw for
+    a specification of the Chromium Event JSON format.
+    """
+
+    @staticmethod
+    def log_event_start(
+        event_name: str,
+        time_ns: int,
+        metadata: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """
+        Logs the start of a single event.
+        :param str event_name Name of event to appear in trace
+        :param time_ns Timestamp in nanoseconds
+        :param metadata: Any extra metadata associated with this event
+        """
+        ChromiumEventLogger._log_timed_event(
+            event_name,
+            time_ns,
+            "B",
+            metadata,
+        )
+
+    @staticmethod
+    def log_event_end(
+        event_name: str,
+        time_ns: int,
+        metadata: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """
+        Logs the end of a single event. This function should only be
+        called after log_event_start with the same event_name.
+        :param event_name: Name of event to appear in trace
+        :param time_ns: Timestamp in nanoseconds
+        :param metadata: Any extra metadata associated with this event
+        """
+        ChromiumEventLogger._log_timed_event(
+            event_name,
+            time_ns,
+            "E",
+            metadata,
+        )
+
+    @staticmethod
+    def _log_timed_event(
+        event_name: str,
+        time_ns: int,
+        phase: str,
+        metadata: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        """
+        Logs a timed event in chromium format. See log_event_start, log_event_end, etc.
+        """
+        event = {
+            "name": event_name,
+            "ts": time_ns / 1000,  # Chromium events are in ms
+            "args": metadata,
+            "ph": phase,
+            "pid": 0,  # pid should be specified on all logs, we don't personally care about the actual process id
+        }
+        torch._logging.trace_structured(
+            "chromium_event",
+            payload_fn=lambda: event,
```

Comment:
It's kind of interesting you decided to put this in payload rather than metadata. Metadata would have been a more obvious place to put it to me, since these records are small.

Reply:
I think I wasn't sure if the records would always stay small: for now, the biggest thing we're logging is `debug_lines` for instant events (unless you want a different structure for timed vs. instant events), but I don't think there's a guarantee that we'll always have small payloads.